src/crimson/osd/heartbeat.cc

```diff
@@ -304,6 +304,29 @@ seastar::future<> Heartbeat::maybe_share_osdmap(
 {
   const osd_id_t from = m->get_source().num();
   const epoch_t osdmap_epoch = service.get_map()->get_epoch();
+  const epoch_t peer_epoch = m->map_epoch;
+  auto found = peers.find(from);
+  if (found == peers.end()) {
+    return seastar::now();
+  }
+  auto& peer = found->second;
+
+  if (peer_epoch > peer.get_last_epoch_sent()) {
+    logger().debug("{} updating session's last epoch sent "
+                   "from {} to peer's (id: {}) map epoch of {}",
+                   __func__, peer.get_last_epoch_sent(),
+                   from, peer_epoch);
+    peer.set_last_epoch_sent(peer_epoch);
+  }
```

Comment:
in what situation peer_epoch < peer.get_last_epoch_sent()? that means peer's epoch decreased?

Reply:
In the situation where we need to update the the peer's `last_epoch_sent()`. It doesn't mean the epoch decreases, its there to keep the `last_epoch_sent()` updated.

Reply:
I compared with OSDService::maybe_share_map(), and seems they are a little different. 
for classic, for example, if peer_epoch =4, peer.last_epoch_sent = 3, will set peer.last_epoch_sent =4, if osdmap_epoch = 6, will do send_incremental_map(epoch 4 to 6) and set peer.last_epoch_sent = 6, if next ping peer_epoch = 4(epcoh not updated on peer but next ping send), and osdmap_epoch = 8, will send_incremental_map(epoch 6 to 8) not from (epoch 4 to 8),
here, the PR logic always compare  if (osdmap_epoch > m->map_epoch) { and redundant send map(epoch 4 to 8).  if we always compare osdmap_epoch > m->map_epoch, and send redundant information,  needn't do the two new condition check. 
so: 1. remove next if (osdmap_epoch > m->map_epoch) {,
2. set peer.last_epoch_sent = osdmap_epoch, after send_incremental_map,
3. and service.send_incremental_map_to_osd(form, peer.last_epoch_sent); 
then the logic should be same. 
could you please go through the OSDService::maybe_share_map() and compare the logic, am I right?

Reply:
Seems correct. I'll push a follow up PR, thank you!